cl/redisapp/leaderfollower/leaderfollower.go

```diff
@@ -0,0 +1,314 @@
+	lfm.leaderProc.Start()
+
+	defer func() {
+		err := lfm.leaderProc.Stop()
+		if err != nil {
+			lfm.logger.Error("Error stopping leader election", "error", err)
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			lfm.logger.Debug("Leader election handler exiting")
+			return
+		case <-lfm.promotedCh:
+			lfm.isLeader.Store(true)
+			lfm.logger.Info("Node promoted to leader")
+		case <-lfm.demotedCh:
+			lfm.isLeader.Store(false)
+			lfm.logger.Info("Node demoted from leader")
+		case err := <-lfm.erroredCh:
+			lfm.logger.Error("Leader election error", "error", err)
+		}
+	}
+}
+
+func (lfm *LeaderFollowerManager) Start(ctx context.Context) {
+	lfm.wg.Add(2)
+
+	go func() {
+		defer lfm.wg.Done()
+		lfm.handleLeaderElection(ctx)
+	}()
+
+	go func() {
+		defer lfm.wg.Done()
+		lfm.run(ctx)
+	}()
+}
+
+func (lfm *LeaderFollowerManager) run(ctx context.Context) {
+	lfm.logger.Info("LeaderFollowerManager started")
+	for {
+		select {
+		case <-ctx.Done():
+			lfm.logger.Debug("LeaderFollowerManager exiting")
+			return
+		default:
+			lfm.logger.Info("LeaderFollowerManager running")
+			if lfm.isLeader.Load() {
+				lfm.isFollowerInitialized.Store(false)
+				lfm.logger.Info("Leader: Starting leader work")
+				if err := lfm.leaderWork(ctx); err != nil {
+					lfm.logger.Error("Error in leader work", "error", err)
+					if errors.Is(err, util.ErrFailedAfterNAttempts) {
+						lfm.logger.Error("Leader: failed to reach geth node after max attempts, exiting")
+						err := lfm.leaderProc.Stop()
+						if err != nil {
+							lfm.logger.Error("Leader: Failed to stop leader election", "error", err)
+						}
+						lfm.blockBuilder.LastCallTime = time.Time{}
+					}
+				}
+			} else {
+				if !lfm.isFollowerInitialized.Load() {
+					if err := lfm.blockBuilder.ProcessLastPayload(ctx); err != nil {
+						lfm.logger.Error("Error processing last payload", "error", err)
+						continue
+					}
+					lfm.isFollowerInitialized.Store(true)
+				}
+				if err := lfm.followerWork(ctx); err != nil {
+					lfm.logger.Error("Error in follower work", "error", err)
+				}
+
+				// will do nothing, in case election already started
+				// needed, in case leader stopped being a leader bcs of geth issue
+				lfm.leaderProc.Start()
+			}
+		}
+	}
+}
+
+func (lfm *LeaderFollowerManager) leaderWork(ctx context.Context) error {
+	lfm.logger.Info("Leader: Performing leader tasks")
+
+	// Ensure state is synchronized before starting leader tasks
+	isSync := lfm.HaveMessagesToProcess(ctx)
+
+	// if not sync, return to wait for the signal to be a leader
+	if !isSync {
+		lfm.logger.Info("Leader: State is not synchronized, waiting for follower to catch up")
+		lfm.leaderProc.Stop()
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			lfm.logger.Debug("Leader: Exiting")
+			return nil
+		default:
+			bbState := lfm.stateManager.GetBlockBuildState(ctx)
+			currentStep := bbState.CurrentStep
+
+			switch currentStep {
```

Comment:
nit: the redundant calls to ResetBlockState can be removed by having an anonymous function around the switch case:

```
err := func() error {
  switch...
}()
if err != nil {
  err = errors.Join(err, lfm.ResetBlockState())
}
```
For the default cast you can return a new error so that this is executed.

Reply:
I still have to keep those calls, as when error is happening during `StepFinalizeBlock`, I don't want to reset a state.